core/src/SYCL/Kokkos_SYCL_Team.hpp

```diff
@@ -129,10 +135,49 @@ class SYCLTeamMember {
   template <typename ReducerType>
   KOKKOS_INLINE_FUNCTION
       typename std::enable_if<is_reducer<ReducerType>::value>::type
-      team_reduce(ReducerType const& /*reducer*/,
-                  typename ReducerType::value_type& /*value*/) const noexcept {
-    // FIXME_SYCL
-    Kokkos::abort("Not implemented!");
+      team_reduce(ReducerType const& reducer,
+                  typename ReducerType::value_type& value) const noexcept {
+    using value_type = typename ReducerType::value_type;
+
+    // We need to chunk up the whole reduction because we might not have
+    // allocated enough memory.
+    const int maximum_work_range =
+        std::min<int>(m_team_reduce_size / sizeof(value_type), team_size());
+
+    int smaller_power_of_two = 1;
+    while ((smaller_power_of_two << 1) < maximum_work_range)
+      smaller_power_of_two <<= 1;
+
+    // We need an intermediate array since value might be identical to
+    // reducer.reference and might be shared.
+    value_type intermediate;
+
+    const int idx        = team_rank();
+    auto reduction_array = static_cast<value_type*>(m_team_reduce);
+
+    // Load values into the first maximum_work_range values of the reduction
+    // array in chunks. This means that only threads with an id in the
+    // corresponding chunk load values and the reduction is always done by the
+    // first smaller_power_of_two threads.
+    for (int start = 0; start < team_size(); start += maximum_work_range) {
```

Comment:
Make this loop just around the joining into the scratch space. So you only have to do the reduction tree once, i.e. first everyone joins (by chunk) into the scratch array. And then you do one tree reduction.

Reply:
Thanks, @crtrott! That's a good idea!